### PR TITLE
Refactor wasm into arch ##arch

### PR DESCRIPTION
**Pull request by swoops**
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

Consider this a work in progress, though you could merge if you wanted. There are still a parsing bugs that should be worked out.

### CODE REVIEW

File: libr/arch/p/wasm/wasm.c [hunk 1]
```diff
@@ -468,16 +468,12 @@ R_IPI int wasm_asm(const char *str, ut8 *buf, int buf_len) {
 			}
 		}
 	}
-	// Abort
-	if (len == 0) goto err;
 	// TODO: parse immediates
-	return len;
-  err:
-	return -1;
+	return len? len: -1;
```
Comment (trufae) [inline libr/arch/p/wasm/wasm.c:472]:
```suggestion
	return len > 0? len: -1;
```

File: libr/bin/p/bin_wasm.c [hunk 2]
```diff
@@ -430,10 +430,44 @@ static int get_fcn_offset_from_id(RBinFile *bf, int ordinal) {
 	return -1;
 }
 
+static int _code_frm_addr(const void *_code, const void *_needle) {
+	int addr = *(int *)_needle;
+	const RBinWasmCodeEntry *code = _code;
+	if (addr < code->code) {
+		return 1;
+	} else if (addr >= code->code + code->len) {
```
Comment (trufae) [inline libr/bin/p/bin_wasm.c:438]: dont else after return

File: libr/bin/p/bin_wasm.c [hunk 3]
```diff
@@ -430,10 +430,44 @@ static int get_fcn_offset_from_id(RBinFile *bf, int ordinal) {
...
+	RBinWasmObj *bin = R_UNWRAP3 (bf, o, bin_obj);
+	if (bin) {
+		RPVector *codes = r_bin_wasm_get_codes (bin);
+		if (codes) {
+			int n = r_pvector_bsearch (codes, (void *)&addr, _code_frm_addr);
+			RBinWasmCodeEntry *code = vector_at (codes, n);
+			if (code) {
+				if (start) {
+					return code->code;
+				} else {
```
Comment (trufae) [inline libr/bin/p/bin_wasm.c:454]:
```suggestion
				}
```

Comment (trufae) [reply libr/bin/p/bin_wasm.c:454]: dont else after return

File: test/db/formats/web_assembly [hunk 4]
```diff
@@ -85,16 +85,16 @@ CMDS=aa;afl
...
 0x00000207    1      3 entry0
 0x00000516    1     19 sym.emscripten_stack_init
-0x0000023d    3    209 sym.main
-0x0000032d    1    126 sym.atoi_via_import
-0x000003b9    1     54 sym.static_int_sum
-0x000003f3    3     14 fcn.000003f3
+0x0000023d    4    209 sym.main
+0x0000032d    2    126 sym.atoi_via_import
+0x000003b9    2     54 sym.static_int_sum
+0x000003f1    3     15 sym._start
```
Comment (trufae) [inline test/db/formats/web_assembly:91]: <3

File: libr/anal/meson.build [hunk 5]
```diff
@@ -112,15 +112,14 @@ r_anal_sources = [
   'p/anal_tricore.c',
   'p/anal_vax.c',
   'arch/vax/vax-dis.c',
-  'p/anal_wasm.c',
+  'p/anal_ws.c',
```
Comment (trufae) [inline libr/anal/meson.build:115]: uh? is this branch rebased?

File: libr/arch/p/wasm/wasm.c [hunk 6]
```diff
@@ -658,15 +654,12 @@ R_IPI int wasm_dis(WasmOp *op, const unsigned char *buf, int buf_len) {
...
 		case WASM_OP_CALL:
-			{
-				ut32 val = 0;
-				size_t n = read_u32_leb128 (buf + 1, buf + buf_len, &val);
-				if (!(n > 0 && n < buf_len)) {
-					goto err;
-				}
-				r_strbuf_setf (sb, "%s %d", opdef->txt, val);
-				op->len += n;
+			size_t n = read_u32_leb128 (buf + 1, buf + buf_len, &op->val);
```
Comment (trufae) [inline libr/arch/p/wasm/wasm.c:657]:
```suggestion
			{
			size_t n = read_u32_leb128 (buf + 1, buf + buf_len, &op->val);
```

File: libr/arch/p/wasm/wasm.c [hunk 7]
```diff
@@ -658,15 +654,12 @@ R_IPI int wasm_dis(WasmOp *op, const unsigned char *buf, int buf_len) {
...
+			if (n <= 0 || n >= buf_len) {
+				goto err;
 			}
+			r_strbuf_setf (sb, "%s %d", opdef->txt, op->val);
+			op->len += n;
```
Comment (trufae) [inline libr/arch/p/wasm/wasm.c:662]:
```suggestion
			op->len += n;
			}
```

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/anal/meson.build (hunk 5 above)
Comment (swoops) [reply libr/anal/meson.build:115]: yeah... I must of messed that up on merge conflicts. Thanks for checking it. I will get it fixed.

File: test/db/formats/web_assembly (hunk 4 above)
Comment (swoops) [reply test/db/formats/web_assembly:91]: Yeah, fixed a couple things there, control flow and function sizes should be a bit more wasm like.

**Comment by swoops** [comment]
^^ This last push saves on unnecessary allocations. I don't think it will count as an ABI change, if so I can separate it or something.

**Comment by trufae** [comment]
Pls rebase again and check for the anal_plugins and arch_plugins arrays in libr/meson.build

**Pull request merged**